python-package/xgboost/spark/model.py

```diff
@@ -1,28 +1,73 @@
-# type: ignore
 """Xgboost pyspark integration submodule for model API."""
 # pylint: disable=fixme, invalid-name, protected-access, too-few-public-methods
 import base64
+import logging
 import os
 import uuid
-
-from pyspark import SparkFiles, cloudpickle
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Dict,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from pyspark import SparkContext, SparkFiles, cloudpickle
 from pyspark.ml.util import DefaultParamsReader, DefaultParamsWriter, MLReader, MLWriter
 from pyspark.sql import SparkSession
 
 from xgboost.core import Booster
 
 from .utils import get_class_name, get_logger
 
-
-def _get_or_create_tmp_dir():
+if TYPE_CHECKING:
+    from .. import XGBClassifier, XGBRanker, XGBRegressor
+    from . import (
+        SparkXGBClassifier,
+        SparkXGBClassifierModel,
+        SparkXGBRanker,
+        SparkXGBRankerModel,
+        SparkXGBRegressor,
+        SparkXGBRegressorModel,
+    )
+
+    # xgboost types
+    XGB_ESTIMATOR = Union[XGBClassifier, XGBRanker, XGBRegressor]
+
+    # pyspark estimator/model types
+    SPARK_XGB_ESTIMATOR = Union[SparkXGBClassifier, SparkXGBRanker, SparkXGBRegressor]
+    SPARK_XGB_ESTIMATOR_TYPE = Union[
+        Type[SparkXGBClassifier], Type[SparkXGBRanker], Type[SparkXGBRegressor]
+    ]
+    SPARK_XGB_MODEL = Union[
+        SparkXGBClassifierModel, SparkXGBRegressorModel, SparkXGBRankerModel
+    ]
+    SPARK_XGB_MODEL_TYPE = Union[
+        Type[SparkXGBClassifierModel],
+        Type[SparkXGBRegressorModel],
+        Type[SparkXGBRankerModel],
+    ]
+
+    SPARK_XGB_INSTANCE = TypeVar(
+        "SPARK_XGB_INSTANCE", SPARK_XGB_ESTIMATOR, SPARK_XGB_MODEL
+    )
```

Comment:
- Please consider the inheritance structure here instead of using Union.
- Please use `CamelCase` for types.

Reply:
@trivialfis any thoughts on that?

Reply:
I think it's the same issue one would encounter for any typed language, the modules need to be written in a structured way. Python duck typing hides the issue, now that we want to annotate the code, some refactoring needs to be done.